🌊 Add type safety to Painless conditions

simianhacker opened the pull request:
## 🍒  Summary

This PR closes https://github.com/elastic/streams-program/issues/18 by adding some basic type checking to the painless output for the Stream conditions. The new code will check to ensure that none of the fields used in the condition are `Map` objects. Then it wraps the if statement in a `try/catch`.

### Condition
```Typescript
{
  and: [
    { field: 'log.logger', operator: 'eq' as const, value: 'nginx_proxy' },
    {
      or: [
        { field: 'log.level', operator: 'eq' as const, value: 'error' },
        { field: 'log.level', operator: 'eq' as const, value: 'ERROR' },
      ],
    },
  ],
}
```

### Before

```
(ctx.log?.logger !== null && ctx.log?.logger == "nginx_proxy") && ((ctx.log?.level !== null && ctx.log?.level == "error") || (ctx.log?.level !== null && ctx.log?.level == "ERROR"))
```

### After

```
if (ctx.log?.logger instanceof Map || ctx.log?.level instanceof Map) {
  return false;
}
try {
  if ((ctx.log?.logger !== null && ctx.log?.logger == "nginx_proxy") && ((ctx.log?.level !== null && ctx.log?.level == "error") || (ctx.log?.level !== null && ctx.log?.level == "ERROR"))) {
    return true;
  }
  return false;
} catch (Exception e) {
  return false;
}
```


<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!

flash1293 commented:
Thanks for setting this up, @simianhacker  - works really well, the only thing I think would make sense to add is to cast to number types when doing gt, gte and so on both for values and fields.

Otherwise this situation will not work as I would expect:

```
// condition
{
            "field": "numberfield",
            "operator": "gt",
            "value": "15"
          }

// Doc (will not be rerouted)
POST logs/_doc
{
  "numberfield": "18"
}
```

```
// condition
{
            "field": "numberfield",
            "operator": "gt",
            "value": 15
          }

// Doc (will not be rerouted)
POST logs/_doc
{
  "numberfield": "18"
}
```


```
// condition
{
            "field": "numberfield",
            "operator": "gt",
            "value": "15"
          }

// Doc (will not be rerouted)
POST logs/_doc
{
  "numberfield": 18
}
```

simianhacker commented:
@flash1293 I added some type safety to `lt`, `lte`, `gt`, `gte`, `contains`, `startsWith`, and `endsWith`. I also added some integration tests. Let me know what you think.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/256715)
* Commit: e392b14e68b4c76d60f73f0f1250ca167fc9ab55

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/256715#01938f7a-641d-49d8-9c49-516b57505f17)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/256715","number":256715,"commit":"e392b14e68b4c76d60f73f0f1250ca167fc9ab55"}],"number":256715}
buildkite-pr-comment-->

flash1293 commented:
Thanks for the changes @simianhacker , this works great, I just found one more case we should handle which is eq/neq:

```
// condition
{
            "field": "numberfield",
            "operator": "gt",
            "value": 15
          }

// Doc (will be rerouted)
POST logs/_doc
{
  "numberfield": 15
}

// Doc (will not be rerouted)
POST logs/_doc
{
  "numberfield": "15"
}

```

simianhacker commented:
@flash1293 I wonder if we shouldn't keep the `eq` and `neq` as EXACT match instead of  converting it to a string. Maybe add `regex` and the caveat for using regex is that it's always converted to a string and then does the `regex` match.

flash1293 commented:
@simianhacker how would the user use eq/neq in the UI?

This is how it looks today:
<img width="430" alt="Screenshot 2024-12-04 at 19 16 56" src="https://github.com/user-attachments/assets/deb8bc05-c56d-46ef-a7ac-ef5075b8a716">

Would the user need to enter `"15"` in order to match 15-the-string? That seems weird to me, especially because it's hard to tell how the data will arrive. Reaching to regex for this also seems weird.

This is not just a theoretical concern, it's pretty common for numbers to be sent as strings instead, but it's also not always the case.

I'm happy to solve this general concern in a different way, but I think we need to solve it somehow - casts just seem like the easiest way to do so. Maybe we need a "like" operator? :)

simianhacker commented:
@flash1293 I see you point, that's kind of a limitation of HTML forms is that it's really hard to discern a number vs a string when the API can accept both. Base on our Slack conversation, let's move towards "ease of use".  The distinction in the API of `"15" vs 15` feels like a "foot gun" for new users when it comes to `eq` and `neq`, they probably mean the same thing.

I'll add the same checks/conversion to `eq` and `neq`

flash1293 commented:
> The distinction in the API of "15" vs 15 feels like a "foot gun" for new users when it comes to eq and neq, they probably mean the same thing

Should we follow through and only allow strings on the API?

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/257109)
* Commit: 9b65b9eeafcc29e262ab523c389b25d084fe5a3a

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/257109#01939340-223a-4947-bbdb-74ecbebdcc3c)

### History
* :broken_heart: [Build #256715](https://buildkite.com/elastic/kibana-pull-request/builds/256715) failed e392b14e68b4c76d60f73f0f1250ca167fc9ab55

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/257109","number":257109,"commit":"9b65b9eeafcc29e262ab523c389b25d084fe5a3a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/256715","number":256715,"commit":"e392b14e68b4c76d60f73f0f1250ca167fc9ab55"}],"number":257109}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/257187)
* Commit: adf8960deeac59011a28b45c84474f6830422b49

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #257109](https://buildkite.com/elastic/kibana-pull-request/builds/257109) failed 9b65b9eeafcc29e262ab523c389b25d084fe5a3a
* :broken_heart: [Build #256715](https://buildkite.com/elastic/kibana-pull-request/builds/256715) failed e392b14e68b4c76d60f73f0f1250ca167fc9ab55

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/257187","number":257187,"commit":"adf8960deeac59011a28b45c84474f6830422b49"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/257109","number":257109,"commit":"9b65b9eeafcc29e262ab523c389b25d084fe5a3a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/256715","number":256715,"commit":"e392b14e68b4c76d60f73f0f1250ca167fc9ab55"}],"number":257187}
buildkite-pr-comment-->

flash1293 reviewed (approved):
LGTM, thanks for all the changes!

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/12181236765

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/203104">](https://github.com/elastic/kibana/pull/203104)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
